### ISSUE TITLE
net::ERR_CONNECTION_RESET in browser when sending large JSON body and not consuming

### ISSUE THREAD

**Issue by MaxGabriel** [yesodweb/yesod#1479]
Cross-posted [from Stackoverflow](https://stackoverflow.com/questions/48351199/yesod-neterr-connection-reset-in-browser-when-sending-large-json-body), since after some investigation I think this is a bug.

I'm sending a large JSON body (~3MB) in a production app (it's a base64 encoded file). When I send a JSON body and don't consume it (just `return ()`), Yesod seems to think this worked:

    POST /send-json
      Accept: */*
      Status: 200 OK 0.000098s

But the browser doesn't get a response back and instead gets `net::ERR_CONNECTION_RESET`. I've reproduced this in Chrome, Safari and Firefox.

If I actually consume the body using `requireCheckJsonBody`, then the issue goes away, which makes me think this is a bug.

If this is a bug, I imagine it could be somewhere lower level than Yesod, e.g. warp/wai.

I reproduced this issue in this repo here: https://github.com/MaxGabriel/large-files. It's the yesod-simple template with just 1 additional commit: https://github.com/MaxGabriel/large-files/commit/38addf762feaa5cbff19b84821c7980eec10b160. Instead of sending a file, I just copy a string a bunch of times to create a huge JSON body (a few megabytes).

I couldn't reproduce this issue sending a large binary file in a form, but I think that may have been because the form was consumed by the Yesod app—I haven't checked that bit.

I can't reproduce this issue from `curl`.

Yesod is configured with:

```haskell
maximumContentLength _ _ = Just (10 * 1024 * 1024) -- 10 megabytes
```

so receiving a large request body shouldn't be a problem.

Version Info: https://gist.github.com/MaxGabriel/10e8a999eabcc9cb566c087978f5c3ac

LTS: LTS-Haskell 9.14

**Comment by snoyberg** [comment]
Any chance you could try reproing with just Warp? I can take a stab at that
myself, but it likely won't be till next week.

**Comment by MaxGabriel** [comment]
Yes I can reproduce with just Warp:

```haskell
#!/usr/bin/env stack
-- stack --resolver lts-9.14 script

{-# LANGUAGE OverloadedStrings #-}
module Main where

import Network.HTTP.Types (status200)
import Network.Wai
import Network.Wai.Handler.Warp

main = run 3000 $ \request respond -> do
  putStrLn "returning response"
  _body <- strictRequestBody request -- Comment out this line to trigger the error
  respond $ responseLBS status200 [] "Hello World"
```

I used the following HTML to generate the request:

```html
<head>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
</head>

<body>

<script>
$(function() {
  $("#send-json").on('click', function(event) {
    var string = "";

    for (i = 0; i < 55000; i++) {
      string += "a".repeat(64)
      string += "\n"
    }

    $.ajax({
      url: 'http://localhost:3000/large-json',
      type: 'POST',
      contentType: "application/json",
      data: JSON.stringify({
        bigString: string,
      }),
      success: function (data) {
        console.log(data);
      },
      error: function (data) {
        console.log("Error: " + data);
      },
    });
  });
});
</script>
<button id="send-json">Send Large JSON</button>
</body>
```

(Note: to workaround CORS issues, I initially served the HTML file using `python -m SimpleHTTPServer 3000`, opened localhost:3000 in my browser, killed the python server, then ran the warp server. Ideally I'd just serve the file from Warp but I didn't look into that).

So far I can't reproduce with a `<form>`, something like:

```
<form action="http://localhost:3000/large-file" method="post">
	<input type="file">
	<input type="submit">
</form>
```

given a 6.2MB file gets the "Hello World" response delivered as a new web page.

**Comment by snoyberg** [comment]
I'm not sure that this is a bug on the Warp side. I can explain what's happening. When you don't consume the request body yourself in your application, Warp tries to implement an optimization to avoid excessive bandwidth usage: it will send the response body immediately, then close the connection. You can see this with the following telnet session against the example Warp application you gave above:

```
$ telnet localhost 3000
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
POST / HTTP/1.1
content-length: 3000000

HTTP/1.1 200 OK
Transfer-Encoding: chunked
Date: Tue, 30 Jan 2018 08:56:24 GMT
Server: Warp/3.2.13

000B
Hello World
0

Connection closed by foreign host.
```

Even though I indicated that I had a 3mb request body, Warp did not wait for it, gave the response, and then closed the connection. This seems to work fine with `curl`, and unless I'm mistaken is allowed by the HTTP spec (though I could be wrong). It may be worth changing the behavior of Warp here, or at least providing a setting for it. CC @kazu-yamamoto for input.

**Comment by MaxGabriel** [comment]
Ah, ok. This might be the relevant part of the spec? https://tools.ietf.org/html/rfc7230#section-6.6

Some discussion about this on the nodeJS side: https://github.com/nodejs/node/issues/9085 https://github.com/nodejs/node/issues/12339

This does seem somewhat edge-case, but I could definitely see situations where e.g. the server immediately rejects the request without sending a response because the client was missing certain headers.

It doesn’t affect me so I’m in no hurry about this, either.

In a car so haven’t looked too closely into this since your reply

**Comment by snoyberg** [comment]
Yes, those links look correct. The only other thing I can think of here is that, perhaps, Warp is in violation by not sending a `connection: close` header. However, I tried setting that manually in the response, and it seemed to have no impact on browser behavior.

**Comment by kazu-yamamoto** [comment]
I don't think that I understand this perfectly. But what will happen if we use `shutdown` instead of `close` in Warp when a request body remains?

**Comment by snoyberg** [comment]
I'm not sure how that would change things.

**Comment by kazu-yamamoto** [comment]
If a browser gets an error of connection close while sending data, the browser has no chance to read a response from the server. If the server uses `shutdown`, the browser does not get an error of connection close, then it gets chance to read a response. This might change the situation.

**Comment by MaxGabriel** [comment]
If my understanding is correct, this is what @kazu-yamamoto is referencing from the HTTP spec I linked to:

>>> To avoid the TCP reset problem, servers typically close a connection
   in stages.  First, the server performs a half-close by closing only
   the write side of the read/write connection.  The server then
   continues to read from the connection until it receives a
   corresponding close by the client, or until the server is reasonably
   certain that its own TCP stack has received the client's
   acknowledgement of the packet(s) containing the server's last
   response.  Finally, the server fully closes the connection.

**Comment by MaxGabriel** [comment]
@nh2 Yes, I can confirm that `setMaximumBodyFlush Nothing` fixes it. Updated script using that:

```haskell
#!/usr/bin/env stack
-- stack --resolver lts-9.14 script

{-# LANGUAGE OverloadedStrings #-}
module Main where

import Network.HTTP.Types (status200)
import Network.Wai
import Network.Wai.Handler.Warp

main = runSettings (setMaximumBodyFlush Nothing defaultSettings) $ \request respond -> do
  putStrLn "returning response"
  -- _body <- strictRequestBody request -- This line no longer necessary
  respond $ responseLBS status200 [] "Hello World"
```
